src/encoded/batch_download.py

```diff
@@ -305,28 +304,40 @@ def batch_download(context, request):
         search_params=request.matchdict['search_params']
     )
     files = [metadata_link]
-    if 'files.file_type' in param_list:
-        for exp in results['@graph']:
-            for f in exp.get('files', []):
-                if f['file_type'] in param_list['files.file_type']:
-                    files.append('{host_url}{href}'.format(
+    exp_files_gen = filter_restricted_files(results)
+    for exp_file in exp_files_gen:
+        if 'files.file_type' in param_list:
```

Comment:
This files type check is a filter too, it's just conditional.  Add this to a function also.  Then, in the end we're just building the urls from a generator.  Splitting this up into functions will make testing easier.

Reply:
I was playing around with this and I think we're going about it wrong.  We should make the exp file generator and then run it through our conditions.  This way it's easy to add more filters.
```
    type_check = None
    if 'files.file_type' in param_list:
        type_check = param_list.get('files.file_type', [])
    exp_files_gen = (
        exp_file
        for exp in results['@graph']
        for exp_file in exp['files']
    )
    files = []
    for exp_file in exp_files_gen:
        if exp_file.get('restricted'):
            continue
        elif type_check is not None and exp_file['file_type'] not in type_check:
            continue
        files.append(
            '{host_url}{href}'.format(
                host_url=request.host_url,
                href=exp_file['href'],
            )
        )
    return Response(
        body='\n'.join(files),
        content_disposition='attachment; filename="%s"' % 'files.txt',
        content_type='text/plain',
    )
```